### PR TITLE
feat(fuzz) - add test progress

### LINKED ISSUE #585: feat: add real-time counter for fuzz/invariant runs
dapptools has a counter like the one in the screenshot below so you can track progress in fuzz runs while those fuzz runs are executing. I recently ran a 10k fuzz run test that uses ffi, so it was quite slow and appears as if tests hang. A counter that updates live like the one below would be helpful for monitoring fuzz/invariant run progress and having confidence tests are still running as expected

image

### LINKED ISSUE #3277: Progress on Invariant Tests
### Component

Forge

### Describe the feature you would like

Currently the invariant testing goes blind while running.

It would be nice to have a dynamic progress on-run like Echidna (where it shows a counter of the current runs / total runs).

### Additional context

_No response_

### LINKED ISSUE #4452: Know on which test forge is stuck
### Component

Forge

### Describe the feature you would like

When I'm running some unit tests on foundry I often get stuck on one for some reason (bad fuzzing, slow rpc, ecc). Since I have hundreds of tests, it would be nice to know on which test I'm stuck. Currently this output 
image
is shown only when all the tests in the contracts have passed (or failed). This doesn't let me know which contract I should check.

### Additional context

In case you're wondering why I don't figure out on the fly that my tests are getting stuck is because I'm using watcher mode.

### PR DESCRIPTION
**Pull request by grandizzy**
## Motivation
Closes #585
Closes #3277
Closes #4452  

## Solution
- make use of `indicatif` crate
- add `--show-progress` arg  (to give more time for feedback / OS tests, then make it default and add a `--stream-results` flag for legacy output)
- long running tasks (fuzz and invariant tests) get their own entry / progress bar in test suite progress. Fuzz tests shows `[current / total]` runs; Invariants show `[current / total]` runs and then the `[current shrink / max shrink]` runs  (if case)
- when a test suite is finished it is moved from running tasks to summary section showing the result summary
- when all test suits are completed then detailed results are streamed

At the moment progress bar is flaky when a failure encountered in fuzz tests (works OK for invariants) because proptest prints messages when
- shrink is disabled (we're shrinking outside of proptest)
- failure persisted in file
https://github.com/proptest-rs/proptest/issues/453 for applying info log level for such messages 

TBD: rn the fail fast flag doesn't stop test case execution (rayon par iter continue to run until all tasks are completed), need to find a way to exit early when such 

rec

### CODE REVIEW

**Comment by mattsse** [comment]
> rn the fail fast flag doesn't stop test case execution (rayon par iter continue to run until all tasks are completed), need to find a way to exit early when such

I think we could add a termination check with something like an `Arc<AtomicBool>` to exit early

**Review by mattsse — COMMENTED** [review]
@DaniPopes should be proceed with this given that this is opt in for now and do a larger cleanup for the nested suite par_iter and exit early separately or do this here?

**Review by DaniPopes — CHANGES_REQUESTED** [review]

File: crates/cli/src/utils/cmd.rs [hunk 1]
```diff
@@ -166,6 +166,61 @@ macro_rules! update_progress {
     };
 }
 
+/// Creates progress object and progress bar.
+#[macro_export]
+macro_rules! init_tests_progress {
```
[the remark this answers is not in the document]
Comment (grandizzy) [reply crates/cli/src/utils/cmd.rs]: I removed those macros and moved logic in structs within newly `progress` mod, pls check https://github.com/foundry-rs/foundry/pull/7914/commits/7aaafc8e20d15e1cb6621483f104906bb9d547d3

**Comment by grandizzy** [comment]
@DaniPopes please take a look when you have time, would love to get this merged and take feedback from community. Thank you 🙏

**Review by DaniPopes — APPROVED** [review]

**Comment by DaniPopes** [comment]
needs rebase + clippy

**Pull request merged**